sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -2949,10 +2949,10 @@ object functions {
   def unix_timestamp(s: Column, p: String): Column = withExpr { UnixTimestamp(s.expr, Literal(p)) }
 
   /**
-   * Converts to a timestamp by casting rules to `TimestampType`.
+   * Converts to a timestamp by casting rules to `TimestampType` in the seconds precision.
    *
    * @param s A date, timestamp or string. If a string, the data must be in a format that can be
-   *          cast to a timestamp, such as `yyyy-MM-dd` or `yyyy-MM-dd HH:mm:ss.SSSS`
+   *          cast to a timestamp, such as `yyyy-MM-dd` or `yyyy-MM-dd HH:mm:ss`
```

Comment:
@MaxGekk, sorry if I missed some contexts but do you mind the ones below don't work? Seems working find in Spark 2.4.4:

```
...
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 2.4.4
      /_/
...
scala> sql("select to_timestamp('2011-01-01 12:12:12.4321')").show(false)
+----------------------------------------+
|to_timestamp('2011-01-01 12:12:12.4321')|
+----------------------------------------+
|2011-01-01 12:12:12.4321                |
+----------------------------------------+


scala> sql("select to_timestamp('2011-01-01 12:12:12.4321')").printSchema()
root
 |-- to_timestamp('2011-01-01 12:12:12.4321'): timestamp (nullable = true)
```

Reply:
Right, it works due to your changes https://github.com/apache/spark/pull/17901 . When the `format` parameter is omitted, `to_timestamp` uses another parsing mechanism via `DateTimeUtils.stringToTimestamp` and the result is not truncated to seconds as we do when the `format` is provided. Weird behavior. :-(

Reply:
Since SimpleDateFormat/FastDateFormat cannot parsing in microseconds precision, I would propose to switch `to_timestamp` on the parser introduced by https://github.com/apache/spark/pull/26507 and port back https://github.com/apache/spark/pull/24420 to 2.4. At the moment, the behavior of `to_timestamp` doesn't seem consistent. /cc @cloud-fan WDYT?

Reply:
> Since SimpleDateFormat/FastDateFormat cannot parsing in microseconds precision

It has been there from the very first place. I think it's not clear if we should take it as a bug to backport.

Reply:
Anyway, I think it's better to discuss about this out of this place.